### PR TITLE
Anonymous tracking

### PR DESCRIPTION
**Pull request by dilyand**

### CODE REVIEW

**Comment by dilyand** [comment]
Smoke test:

```
| Ⰱ :> curl -X POST -i http://localhost:12345/com.snowplowanalytics.snowplow/tp2  -d '{}'
HTTP/1.1 200 OK
Set-Cookie: sp=3256f728-5a30-4d24-8431-c4917e8deeba; Expires=Sat, 27 Nov 2021 15:56:25 GMT; Domain=; Path=/; SameSite=None
P3P: policyref="/w3c/p3p.xml", CP="NOI DSP COR NID PSA OUR IND COM NAV STA"
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Server: akka-http/10.1.12
Date: Fri, 27 Nov 2020 15:56:25 GMT
Content-Type: text/plain; charset=UTF-8
Content-Length: 2
```

> d	127.0.0.1
�v
k}o�UTF-8�ssc-2.1.0-local-stdout$,curl/7.54.0@#/com.snowplowanalytics.snowplow/tp2T{}^Timeout-Access: <function1>Host: localhost:12345User-Agent: curl/7.54.0Accept: */*!application/x-www-form-urlencodedh!application/x-www-form-urlencoded�	localhost�$3256f728-5a30-4d24-8431-c4917e8deebaziAiglu:com.snowplowanalytics.snowplow/CollectorPayload/thrift/1-0-0

```
| Ⰱ :> curl -X POST -i http://localhost:12345/com.snowplowanalytics.snowplow/tp2 -H 'SP-Anonymous: *' -d '{}'
HTTP/1.1 200 OK
P3P: policyref="/w3c/p3p.xml", CP="NOI DSP COR NID PSA OUR IND COM NAV STA"
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
Server: akka-http/10.1.12
Date: Fri, 27 Nov 2020 15:58:13 GMT
Content-Type: text/plain; charset=UTF-8
Content-Length: 2
```

> dunknown
�v
m%�UTF-8�ssc-2.1.0-local-stdout$,curl/7.54.0@#/com.snowplowanalytics.snowplow/tp2T{}^Timeout-Access: <function1>Host: localhost:12345User-Agent: curl/7.54.0Accept: */*SP-Anonymous: *!application/x-www-form-urlencodedh!application/x-www-form-urlencoded�	localhost�$917df74d-c1a2-4d33-85ca-528227e77ab6ziAiglu:com.snowplowanalytics.snowplow/CollectorPayload/thrift/1-0-0

**Comment by paulboocock** [comment]
I've given this a run through and it looks good (once I'd realised you can't set cookies via ngrok). 

The `SP-Anonymous` header appears to correctly prevent the `_sp` cookie from being set and also sets the ip address to `unknown` as previously discussed.

There is one additional area I think we might also need to change. If the user already has cookies set (perhaps from a previous visit where the `_sp` cookie was set), then the browser will automatically include these cookies in the request. There could be other cookies containing user identifiers too (which might not be set by Snowplow).
I think that to ensure we can call this "cookieless" we need to remove the "Cookie" header from the request in the collector so the "Cookie" header is never persisted to the Stream. This isn't something that can be done at the tracker level, the browser automatically includes cookies that match the domain.

As an example, below is a payload which is using `SP-Anonymous` but still includes the `_sp` cookie because the browser automatically included it from an earlier session. I think the `Cookie` header should be removed before the payload is written anywhere so we never persist it in anonymous mode. I also think it is safe to remove the entire `Cookie` header.

```

**Comment by dilyand** [comment]
Thanks @paulboocock , I addressed this here: https://github.com/snowplow/stream-collector/blob/feature/ip-cloak/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala#L348-L365

**Comment by paulboocock** [comment]
> Thanks @paulboocock , I addressed this here: https://github.com/snowplow/stream-collector/blob/feature/ip-cloak/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala#L348-L365

Awesome. Tested and works well! This looks good from my point of view.

File: build.sbt [hunk 1]
```diff
@@ -45,15 +45,16 @@ lazy val commonDependencies = Seq(
...
   description := "Scala Stream Collector for Snowplow raw events",
   scalaVersion := "2.12.10",
+  javacOptions := Seq("-source", "11", "-target", "11"),
   resolvers ++= Dependencies.resolutionRepos
 )
 
 lazy val dockerSettings = Seq(
   maintainer in Docker := "Snowplow Analytics Ltd. <support@snowplowanalytics.com>",
-  dockerBaseImage := "snowplow-docker-registry.bintray.io/snowplow/base-debian:0.1.0",
+  dockerBaseImage := "snowplow-docker-registry.bintray.io/snowplow/base-debian:0.1.1",
```
Comment (chuwy) [inline build.sbt:57]: I think `base-debian` is at 0.2.1 already. @oguzhanunlu can you remind us what's the difference?

File: core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala [hunk 2]
```diff
@@ -336,11 +345,24 @@ class CollectorService(
...
+        request.headers.flatMap {
+          case _: `Remote-Address` | _: `Raw-Request-URI` => None
+          case other => Some(other.toString)
+        }
+      case Some(_) =>
+        request.headers.flatMap {
+          case _: `Remote-Address` | _: `Raw-Request-URI` | _: `X-Forwarded-For` | _: `X-Real-Ip` | _: `Cookie` => None
+          case other => Some(other.toString)
+        }
+    }
```
Comment (chuwy) [inline core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala:365]:
```suggestion
  def headers(request: HttpRequest, spAnonymous: Option[String]): Seq[String] =
    request.headers.flatMap {
      case _: `Remote-Address` | _: `Raw-Request-URI` if spAnonymous.isEmpty => None
      case _: `Remote-Address` | _: `Raw-Request-URI` | _: `X-Forwarded-For` | _: `X-Real-Ip` | _: `Cookie` if spAnoymous.isDefined => None
      case other => Some(other.toString)
    }
```

**Review by chuwy — APPROVED** [review]
Looks good overall. Just one thing I don't fully understand (haven't read any proposal and probably my lack of understanding of HTTP), but it seems we never define how `SP-Anonymous` vallue should look like, the only value I see is `*`, is it widely accepted practice? What if somebody wants to pass `SP-Anonymous: false`?

File: build.sbt (hunk 1 above)
Comment (oguzhanunlu) [reply build.sbt:57]: [`snowplow/base-debian:0.2.2`](https://hub.docker.com/layers/snowplow/base-debian/0.2.2/images/sha256-ebe1ed4340aff0c4ae4c2d8fa8092c080b47aef222826ab52eb1f71f8de8cdee?context=explore) is available.

base-debian 0.1.x and 0.2.x images are based on JDK8 and JDK11 respectively. Given that collector targets JDK11 now, base image should be bumped to `0.2.2` here

**Comment by dilyand** [comment]
> Looks good overall. Just one thing I don't fully understand (haven't read any proposal and probably my lack of understanding of HTTP), but it seems we never define how `SP-Anonymous` vallue should look like, the only value I see is `*`, is it widely accepted practice? What if somebody wants to pass `SP-Anonymous: false`?

It's supposed to always be `"*"`: https://snplow.atlassian.net/browse/PE-1023?focusedCommentId=20928

**Comment by paulboocock** [comment]
My initial intention for `SP-Anonymous` is for a value of `'*'` to be synonymous for "anonymise all". In the future we could perhaps have a value of `'ip,cookie'` or something allowing for more anonymisation flexibility. I'd expect in this release, if a value other than `'*'` is specified then we would not be anonymising - this allows for extension of the feature without introducing breaking changes in the future.

**Review by benjben — APPROVED** [review]
Looks great @dilyand :+1: 

Main thing is that I think that `spAnonymous` should be a `Boolean` rather than an `Option[String]`.

Also, if time allows it would be great to add an integration test for this feature. We could do:
1) `sbt docker:publishLocal`
2) `docker run`
3) Perform `curl -v` request and check that `set-cookie` is set
4) Perform `curl -v` request with `SP-Anonymous` header and check that `set-cookie` is not set
5) `docker stop`
We can do this directly from Github Actions

File: core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala [hunk 3]
```diff
@@ -27,9 +26,15 @@ trait CollectorRoute {
 
   private val headers = optionalHeaderValueByName("User-Agent") &
     optionalHeaderValueByName("Referer") &
-    optionalHeaderValueByName("Raw-Request-URI")
-
-  private val extractors = extractHost & extractClientIP & extractRequest
+    optionalHeaderValueByName("Raw-Request-URI") &
+    optionalHeaderValueByName("SP-Anonymous")
+
+  private[scalastream] def extractors(spAnonymous: Option[String]) =
```
[the remark this answers is not in the document]
Comment (dilyand) [reply core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala:32]: The idea is that this header has a value. Currently, this is always `*` and we don't care about it. But in future, we might want to do things based on the value. So we have multiple options: no header (= no value, or `None`), header with value 1 (= `Some(1)`), header with value 2 (= `Some(2)`), etc.

File: core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala [hunk 4]
```diff
@@ -67,7 +72,8 @@ trait CollectorRoute {
                       request,
                       pixelExpected = false,
                       doNotTrack    = dnt,
-                      Some(ct)
+                      Some(ct),
+                      spAnonymous
```
[the remark this answers is not in the document]
Comment (dilyand) [reply core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala:76]: If the request is anonymous, then we do not want to set a cookie from the collector. This function is actually in charge of the entire response, not just the `Set-Cookie` portion, so `spAnonymous` here is a flag indicating whether the `Set-Cookie` header should be set or not.

File: core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala [hunk 5]
```diff
@@ -27,9 +26,15 @@ trait CollectorRoute {
...
-    optionalHeaderValueByName("Raw-Request-URI")
-
-  private val extractors = extractHost & extractClientIP & extractRequest
+    optionalHeaderValueByName("Raw-Request-URI") &
+    optionalHeaderValueByName("SP-Anonymous")
+
+  private[scalastream] def extractors(spAnonymous: Option[String]) =
+    spAnonymous match {
+      case Some(_) =>
+        extractHost & extractClientIP.map[RemoteAddress](_ => RemoteAddress.Unknown) & extractRequest
```
[the remark this answers is not in the document]
Comment (dilyand) [reply core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala:35]: I've ran a `spirit-walk` test with `SP-Anonymous` and all the events were loaded in BQ with `unknown` IP address. I think we're good.

File: core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjben) [reply core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorRoute.scala:32]: `Currently, this is always`, `But in future, we might` => https://twitter.com/alexelcu/status/1334422126762352643

**Pull request merged**